### ISSUE TITLE
Physik Update!

### ISSUE THREAD

**Issue by Dev-MarkoF** [JirkaDellOro/FUDGE#267]
Nachdem ich Oimo geforked und um Formbasierte Trigger erweitert habe, konnte ich auch die restlichen Verbesserungen die mit einer neueren Oimo Version gekommen sind auf Fudge Übertragen.

**Daher Commit auf Development Branch bitte komplett übernehmen!**

### Änderungen:
- **Formgenaue Trigger**, Oimo Physics löst jetzt Trigger Events aus (Enter/Exit), wenn ein Kontakt entsteht, analog zu "echten" Kontakten, allerdings werden diese Kontakte nicht im Solver berücksichtigt, wie für Trigger üblich. 
E**twas performance sparender (da Event gesteuert, statt Framebasiert) werden jetzt in FUDGE Trigger Events ausgelöst**, die Verwendung des Events entspricht dem vorherigen, nur wird jetzt z.B. eine Sphere auch als solche behandelt.
**Breaking Change!** Trigger sind jetzt keine eigene Kollisionsgruppe mehr, jedes Objekt kann ein Trigger sein, d.h. es ist nun möglich z.B. Trigger zu erstellen die nur für `PHYSICS_GROUP.DEFAULT` oder `1` etc. gültig sind, mehr Optionen. 
Um ein Objekt als Trigger zu markieren, muss jetzt nur die flag `isTrigger = true` gesetzt werden.
D.h. alle Games etc. müssen 1-2 Zeilen pro Trigger angepasst werden.

- **ComponentJointWelding**, eine neue Gelenk Komponente um physische Objekte mit einander zu verbinden, in einer quasi Parent/Child Hierarchie analog zur Transformationswelt. Sie werden zusammengeschweißt, d.h. die Physik behandelt sie als ein Objekt, somit können mehrere Convexe Kollider zu einem Concaven verbunden werden. **Siehe neues Experiment/Marko -> Chair.** 
`(Performance Einfluss gering, aber vorhanden, es konnten im Test bei 500 Joints, a 700 Rb's immer noch 15 Frames gemessen werden)`

- **Einige kleine Verbesserungen**, da der Fork auf **Oimo 1.2.2** anstatt **1.2.0** basiert, d.h. kleinere Filesize, stabilere Joints (siehe Experiment JointCar unfinished), Fehler Behebungen basierend auf open Issues von Oimo etc.

- Alle Tutorials (Text und Code)/ Experimente auf die neusten Fudge Standards / Physik Änderungen überprüft/angepasst -> Tests erfolgreich.

- **Falls die Physik berechnet wird mit einer Frametime von weniger als 30 FPS, wird berechnet, als sei die Zeit von 30 FPS** vergangen, d.h. es wechselt von dynamisch zu konstant und vermindert Bugs bei Frameeinbrüchen. Hinweis! Es ist in der Welt der Physik immer sinnvoller mit konstanter Deltazeit zu berechnen, anstatt `Game.Frametime.`

Auf der Liste für Verbesserungen die ich vor habe steht nur noch ein **Heightmap** Collider, mit diesem ist nicht zeitnah zu rechnen. OimoPhysics selbst kann jederzeit ein Update erhalten, aber der Entwickler hält sich da bewusst bedeckt.

Probleme, gerne bei mir melden, ich bemühe mich um eine zeitnahe Lösung, da Oimo in Haxe geschrieben ist, ist evtl. eine kurze Recherchezeit meinerseits nötig, ich bitte daher um Geduld.

**Comment by JirkaDellOro** [comment]
So... ich komme jetzt dazu, mir das mal anzuschauen. Das klingt ja alles klasse! Dein Commit ist ja direkt in den development-branch gegangen, da muss ich also nix machen, oder? Ich übertrage das demnächst alles auf Master, bevor die größeren Änderungen am Editor kommen. Mit der Heightmap kann man getrost warten, die muss als Komponente erstmal im Editor funktionieren.

**Comment by JirkaDellOro** [comment]
Mir scheint, dass die Trigger nur auf dynamische Collider reagieren. Ist das richtig beobachtet und so gedacht? Ich habe Anwendungsfälle, wo z.B. ein Trigger an meinem Avatar melden soll, was sich in seiner Nähe befindet. Da wäre auch die Kollision mit statischen und kinematischen Collidern sinnvoll.

**Comment by Dev-MarkoF** [comment]
Muss ich nochmal prüfen, sollte aber auch mit statischen Collidern gehen, da diese ebenso Kontakte auslösen.
Das Event funktioniert identisch mit dem bisher bestehenden Collision Event und das sollte auch bei statischen Collidern ausgelöst werden.

In meinem Test für Events sind Trigger auch statisch platziert, diese reagieren auch korrekt.
Also Trigger (statisch) -> Triggered (dynamisch) funktioniert problemlos, andersherum habe ich's zugegebenermaßen nicht probiert, aber ich schau mal rein.

**Comment by JirkaDellOro** [comment]
> Body (dynamic) -> attached Trigger (kinematic) -> Triggered (any State) scheint nicht zu funktionieren.   

Dem scheint zu widersprechen, dass dynamische Bodies den Trigger auslösen. Siehe https://github.com/JirkaDellOro/Prima/tree/master/L05_PhysicsGame

**Comment by Dev-MarkoF** [comment]
Stimmt, danke, ging mit attached Trigger -> Triggered (kinematic/static) nicht, dachte geht dann auch mit dynamic nicht.
Verstärkt meine Theorie, dass der Branch in dem Kinematic/Static sind verworfen wird, bevor das CollisionManifold initial geupdated wird.
Wenn der kinematic Trigger nicht attached ist, dann funktioniert es wie gewünscht, das Problem muss im attached liegen, das anhängen wird aber in Fudge gemacht, nicht Oimo, Oimo übernimmt nur die Werte aus Fudge am Ende des Renderings.

Aber vlt. liegt es genau daran, danke für den Hinweis ich schau mal rein.

**Comment by Dev-MarkoF** [comment]
Danke für die Meldung des Fehlers, die Eventualität Dynamic Body + Kinematic Trigger hab ich nicht getestet.

Ich konnte den Fehler beheben, wie vermutet, handelt es sich um eine Vereinfachung, die Oimo da trifft und einfach gar keine Tests für den Fall durchführt und die direkt aussortiert. Deswegen werden jetzt Trigger generell nicht aussortiert und immer "Kontakte" generiert.

Trigger (dynamic) -> Triggered (dynamic) = ✓
Trigger (kinematic) -> Triggered (dynamic) = ✓
Trigger (static) -> Triggered (kinematic) = ✓
Trigger (dynamic) -> Triggered (static) = ✓
Trigger (kinematic) -> Triggered (kinematic) = ✓
Trigger (kinematic) -> Triggered (static) = ✓

Jetzt sollten alle Fälle abgedeckt sein und hoffentlich durch den Fix keine ungewünschten Nebeneffekte entstanden sein.

Hab den DevBranch aktualisiert, bitte in deinem Kontext testen. Gerne melden, wenn etwas nicht funktioniert.
